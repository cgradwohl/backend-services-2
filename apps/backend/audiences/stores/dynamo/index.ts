import * as dynamoUtil from "~/audiences/util/dynamo";
import { getDocClient } from "~/audiences/lib/doc-client";
import { DocumentClient } from "aws-sdk/clients/dynamodb";

export * as StoreTypes from "./types";
const docClient = getDocClient();

export async function putItem<T>(item: T) {
  const params: DocumentClient.PutItemInput = {
    Item: {
      ...item,
    },
    TableName: dynamoUtil.AUDIENCES_TABLE_NAME,
  };
  await docClient.put(params).promise();
}

export async function updateItem<T>(
  updateQuery: Partial<DocumentClient.UpdateItemInput>
): Promise<T> {
  const { Attributes } = await docClient
    .update({
      ...(updateQuery as DocumentClient.UpdateItemInput),
      ReturnValues: "ALL_NEW",
      TableName: dynamoUtil.AUDIENCES_TABLE_NAME,
    })
    .promise();

  return Attributes as T;
}

export async function getItem<T>(key: DocumentClient.Key): Promise<T> {
  const params = {
    TableName: dynamoUtil.AUDIENCES_TABLE_NAME,
    Key: {
      ...key,
    },
  };
  const response = await docClient.get(params).promise();
  return response.Item as T;
}

export async function deleteItem(key: DocumentClient.Key) {
  const params: DocumentClient.DeleteItemInput = {
    TableName: dynamoUtil.AUDIENCES_TABLE_NAME,
    Key: {
      ...key,
    },
  };
  await docClient.delete(params).promise();
}

export async function query(
  args: DocumentClient.QueryInput
): Promise<DocumentClient.QueryOutput> {
  return docClient.query(args).promise();
}
