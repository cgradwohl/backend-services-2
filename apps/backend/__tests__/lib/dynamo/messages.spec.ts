import * as dynamo from "~/lib/dynamo";
import {
  create,
  incrementErrorCount,
  markClicked,
  markDelivered,
  markEmail,
  markOpened,
  markRead,
  markSent,
  markSimulated,
  markUndeliverable,
  markUndeliverableFromDelivery,
  markUnmapped,
  markUnread,
  setBilledUnits,
  setNotificationId,
} from "~/lib/dynamo/messages";
import { markUnroutable } from "~/lib/dynamo/messages-v3-adapter";
import { shouldMarkUndeliverable } from "~/lib/dynamo/should-mark-undeliverable";
import {
  IMessageHistory,
  IRoutedMessageHistory,
  IUndeliverableMessageHistory,
  MessageHistoryType,
  RoutedMessageHistoryTypes,
} from "~/lib/message-service/types";

jest.mock("~/lib/dynamo", () => {
  return {
    put: jest.fn(),
    query: jest.fn(),
    update: jest.fn(),
  };
});

jest.mock("~/lib/get-environment-variable");

jest.mock("~/lib/get-hash-from-range", () => ({
  getHashFromRange: jest.fn().mockReturnValue(1),
}));

jest.mock("~/providers", () => {
  return {
    providerWithDelivery: {
      deliveryStatusStrategy: "polling",
    },
    providerWithoutDelivery: {},
  };
});

jest.mock("~/lib/capture-exception", () => {
  return jest.fn();
});

describe("lib/messages", () => {
  describe("v3", () => {
    const OLD_ENV = { ...process.env };

    afterAll(() => {
      process.env = { ...OLD_ENV };
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    beforeEach(() => {
      process.env.PREFIX = "jest-mock-table-name";
      process.env.MESSAGES_V3_TABLE = "jest-mock-table-name-messages-v3";

      Date.now = jest.fn(() => 1632426903638);
      jest.resetModules();
    });

    it("should create a new message", async () => {
      await create(
        "MOCK_TENANT_ID",
        "MOCK_EVENT_ID",
        "MOCK_RECIPIENT_ID",
        "MOCK_MESSAGE_ID"
      );
      expect(dynamo.put).toBeCalledTimes(1);
      expect(dynamo.put).toMatchSnapshot();
    });

    it("should increment the error count on a message", async () => {
      await incrementErrorCount("MOCK_TENANT_ID", "MOCK_EVENT_ID");
      expect(dynamo.update).toBeCalledTimes(1);
      expect(dynamo.update).toMatchSnapshot();
    });

    it("should set the billing units id value on a message", async () => {
      await setBilledUnits("MOCK_TENANT_ID", "MOCK_EVENT_ID", 1.0);
      expect(dynamo.update).toBeCalledTimes(1);
      expect(dynamo.update).toMatchSnapshot();
    });

    it("should set the notification id value on a message", async () => {
      await setNotificationId(
        "MOCK_TENANT_ID",
        "MOCK_EVENT_ID",
        "MOCK_NOTIFICATION_ID"
      );
      expect(dynamo.update).toBeCalledTimes(1);
      expect(dynamo.update).toMatchSnapshot();
    });

    it("should set the email address value on a message", async () => {
      await markEmail("MOCK_TENANT_ID", "MOCK_EVENT_ID", "email@example.com");
      expect(dynamo.update).toBeCalledTimes(1);
      expect(dynamo.update).toMatchSnapshot();
    });

    it("should mark a message as unmapped", async () => {
      await markUnmapped("MOCK_TENANT_ID", "MOCK_MESSAGE_ID");
      expect(dynamo.update).toBeCalledTimes(1);
      expect(dynamo.update).toMatchSnapshot();
    });

    describe("markUnroutable", () => {
      const MOCK_LOG_ITEMS = [
        {
          id: "MOCK_EVENT_LOG_ID_1",
          json: {},
          messageId: "MOCK_MESSAGE_ID",
          type: "event:received",
        },
      ];

      beforeEach(() => {
        (dynamo.query as jest.Mock).mockResolvedValue({
          Items: MOCK_LOG_ITEMS,
        });
      });

      [
        { event: "sent", type: "provider:sent" },
        { event: "delivered", type: "provider:delivered" },
        { event: "opened", type: "event:opened" },
        { event: "clicked", type: "event:click" },
      ].forEach(({ event, type }) => {
        it(`should not mark a message unroutable if "${event}"`, async () => {
          (dynamo.query as jest.Mock).mockResolvedValue({
            Items: [
              ...MOCK_LOG_ITEMS,
              {
                id: "MOCK_EVENT_LOG_ID_2",
                json: { channel: { id: "MOCK_CHANNEL", label: "MOCK_LABEL" } },
                messageId: "MOCK_MESSAGE_ID",
                type,
              },
            ],
          });
          await markUnroutable("MOCK_TENANT_ID", "MOCK_MESSAGE_ID");
          expect(dynamo.update).not.toBeCalled();
        });
      });

      it("should mark a message as unroutable", async () => {
        await markUnroutable("MOCK_TENANT_ID", "MOCK_MESSAGE_ID");
        expect(dynamo.update).toBeCalledTimes(1);
        expect(dynamo.update).toMatchSnapshot();
      });
    });

    describe("markUndeliverable", () => {
      const MOCK_LOG_ITEMS = [
        {
          id: "MOCK_EVENT_LOG_ID_1",
          json: {},
          messageId: "MOCK_MESSAGE_ID",
          type: "message:received",
        },
      ];

      beforeEach(() => {
        (dynamo.query as jest.Mock).mockResolvedValue({
          Items: MOCK_LOG_ITEMS,
        });
      });

      [
        { event: "sent", type: "provider:sent" },
        { event: "delivered", type: "provider:delivered" },
        { event: "opened", type: "event:opened" },
        { event: "clicked", type: "event:click" },
      ].forEach(({ event, type }) => {
        it(`should not mark a message undeliverable if "${event}"`, async () => {
          (dynamo.query as jest.Mock).mockResolvedValue({
            Items: [
              ...MOCK_LOG_ITEMS,
              {
                id: "MOCK_EVENT_LOG_ID_2",
                json: { channel: { id: "MOCK_CHANNEL", label: "MOCK_LABEL" } },
                messageId: "MOCK_MESSAGE_ID",
                type,
              },
            ],
          });
          await markUndeliverable("MOCK_TENANT_ID", "MOCK_MESSAGE_ID");
          expect(dynamo.update).not.toBeCalled();
        });
      });

      it("should mark a message as undeliverable", async () => {
        await markUndeliverable("MOCK_TENANT_ID", "MOCK_MESSAGE_ID");
        expect(dynamo.update).toBeCalledTimes(1);
        expect(dynamo.update).toMatchSnapshot();
      });

      it("should mark a message as undeliverable with configuration", async () => {
        await markUndeliverable("MOCK_TENANT_ID", "MOCK_MESSAGE_ID", {
          configuration: "MOCK_CONFIGURATION",
        });
        expect(dynamo.update).toBeCalledTimes(1);
        expect(dynamo.update).toMatchSnapshot();
      });

      it("should mark a message as undeliverable with provider", async () => {
        await markUndeliverable("MOCK_TENANT_ID", "MOCK_MESSAGE_ID", {
          provider: "MOCK_PROVIDER",
        });
        expect(dynamo.update).toBeCalledTimes(1);
        expect(dynamo.update).toMatchSnapshot();
      });

      it("should mark a message as undeliverable with internal courier error", async () => {
        await markUndeliverable("MOCK_TENANT_ID", "MOCK_MESSAGE_ID", {
          errorMessage: "Internal Courier Error",
        });
        expect(dynamo.update).toBeCalledTimes(1);
        expect(dynamo.update).toMatchSnapshot();
      });
    });

    describe("markUndeliverableFromDelivery", () => {
      const MOCK_LOG_ITEMS = [
        {
          id: "MOCK_EVENT_LOG_ID_1",
          json: {},
          messageId: "MOCK_MESSAGE_ID",
          type: "message:received",
        },
      ];

      beforeEach(() => {
        (dynamo.query as jest.Mock).mockResolvedValue({
          Items: MOCK_LOG_ITEMS,
        });
      });

      [
        { event: "delivered", type: "provider:delivered" },
        { event: "opened", type: "event:opened" },
        { event: "clicked", type: "event:click" },
      ].forEach(({ event, type }) => {
        it(`should not mark a message undeliverable if "${event}"`, async () => {
          (dynamo.query as jest.Mock).mockResolvedValue({
            Items: [
              ...MOCK_LOG_ITEMS,
              {
                id: "MOCK_EVENT_LOG_ID_2",
                json: { channel: { id: "MOCK_CHANNEL", label: "MOCK_LABEL" } },
                messageId: "MOCK_MESSAGE_ID",
                type,
              },
            ],
          });
          await markUndeliverableFromDelivery(
            "MOCK_TENANT_ID",
            "MOCK_MESSAGE_ID"
          );
          expect(dynamo.update).not.toBeCalled();
        });
      });

      it("should mark a message as undeliverable", async () => {
        await markUndeliverableFromDelivery(
          "MOCK_TENANT_ID",
          "MOCK_MESSAGE_ID"
        );
        expect(dynamo.update).toBeCalledTimes(1);
        expect(dynamo.update).toMatchSnapshot();
      });

      it("should mark a message as undeliverable with configuration", async () => {
        await markUndeliverableFromDelivery(
          "MOCK_TENANT_ID",
          "MOCK_MESSAGE_ID",
          {
            configuration: "MOCK_CONFIGURATION",
          }
        );
        expect(dynamo.update).toBeCalledTimes(1);
        expect(dynamo.update).toMatchSnapshot();
      });

      it("should mark a message as undeliverable with provider", async () => {
        await markUndeliverableFromDelivery(
          "MOCK_TENANT_ID",
          "MOCK_MESSAGE_ID",
          {
            provider: "MOCK_PROVIDER",
          }
        );
        expect(dynamo.update).toBeCalledTimes(1);
        expect(dynamo.update).toMatchSnapshot();
      });

      it("should mark a message as undeliverable with internal courier error", async () => {
        await markUndeliverableFromDelivery(
          "MOCK_TENANT_ID",
          "MOCK_MESSAGE_ID",
          {
            errorMessage: "Internal Courier Error",
          }
        );
        expect(dynamo.update).toBeCalledTimes(1);
        expect(dynamo.update).toMatchSnapshot();
      });
    });

    it("should mark a message as simulated", async () => {
      await markSimulated(
        "MOCK_TENANT_ID",
        "MOCK_MESSAGE_ID",
        "MOCK_PROVIDER",
        "MOCK_CONFIGURATION"
      );
      expect(dynamo.update).toBeCalledTimes(1);
      expect(dynamo.update).toMatchSnapshot();
    });

    it("should mark a message as sent", async () => {
      await markSent(
        "MOCK_TENANT_ID",
        "MOCK_MESSAGE_ID",
        "MOCK_PROVIDER",
        "MOCK_CONFIGURATION"
      );
      // 2: updateSentExtraData and updateSentStatus
      expect(dynamo.update).toBeCalledTimes(2);
      expect(dynamo.update).toMatchSnapshot();
    });

    it("should mark a message as delivered", async () => {
      await markDelivered(
        "MOCK_TENANT_ID",
        "MOCK_MESSAGE_ID",
        "MOCK_PROVIDER",
        "MOCK_CONFIGURATION"
      );
      expect(dynamo.update).toBeCalledTimes(1);
      expect(dynamo.update).toMatchSnapshot();
    });

    it("should mark a message as opened", async () => {
      const timestamp = 1632427827545;
      await markOpened("MOCK_TENANT_ID", "MOCK_MESSAGE_ID", timestamp);
      expect(dynamo.update).toBeCalledTimes(1);
      expect(dynamo.update).toMatchSnapshot();
    });

    it("should mark a message as clicked", async () => {
      const timestamp = 1632427827545;
      await markClicked("MOCK_TENANT_ID", "MOCK_MESSAGE_ID", timestamp);
      expect(dynamo.update).toBeCalledTimes(1);
      expect(dynamo.update).toMatchSnapshot();
    });

    it("should mark a message as read", async () => {
      const timestamp = 1632427827545;
      await markRead("MOCK_TENANT_ID", "MOCK_MESSAGE_ID");
      expect(dynamo.update).toBeCalledTimes(1);
      expect(dynamo.update).toMatchSnapshot();
    });

    it("should mark a message as unread", async () => {
      const timestamp = 1632427827545;
      await markUnread("MOCK_TENANT_ID", "MOCK_MESSAGE_ID");
      expect(dynamo.update).toBeCalledTimes(1);
      expect(dynamo.update).toMatchSnapshot();
    });
  });

  describe("when seeing if a message should be marked undeliverable", () => {
    it("will return false if history contains CLICKED", () => {
      const history: Array<IRoutedMessageHistory<RoutedMessageHistoryTypes>> = [
        {
          channel: {
            id: "ab17b3cd-21a7-47fb-b1dc-a60ee7f6c2f4",
            label: "",
          },
          integration: {
            id: "4276e969-9c45-42de-9cc1-7175623d4d27",
            provider: "providerWithDelivery",
          },
          ts: 16134123412,
          type: "CLICKED",
        },
      ];

      expect(shouldMarkUndeliverable(history, ["CLICKED"])).toBe(false);
    });

    it("will return false if history contains DELIVERED", () => {
      const history: Array<IRoutedMessageHistory<RoutedMessageHistoryTypes>> = [
        {
          channel: {
            id: "ab17b3cd-21a7-47fb-b1dc-a60ee7f6c2f4",
            label: "",
          },
          integration: {
            id: "4276e969-9c45-42de-9cc1-7175623d4d27",
            provider: "providerWithDelivery",
          },
          ts: 16134123412,
          type: "DELIVERED",
        },
      ];

      expect(shouldMarkUndeliverable(history, ["DELIVERED"])).toBe(false);
    });

    it("will return false if history contains OPENED", () => {
      const history: Array<IRoutedMessageHistory<RoutedMessageHistoryTypes>> = [
        {
          channel: {
            id: "ab17b3cd-21a7-47fb-b1dc-a60ee7f6c2f4",
            label: "",
          },
          integration: {
            id: "4276e969-9c45-42de-9cc1-7175623d4d27",
            provider: "providerWithDelivery",
          },
          ts: 16134123412,
          type: "OPENED",
        },
      ];

      expect(shouldMarkUndeliverable(history, ["OPENED"])).toBe(false);
    });

    it("will return true if history contains no routed events", () => {
      const history: Array<IMessageHistory<MessageHistoryType>> = [
        {
          ts: 1610475430221,
          type: "ENQUEUED",
        },
      ];

      expect(shouldMarkUndeliverable(history, [])).toBe(true);
    });

    it("will not throw an error if history contains undeliverable w/o routing details", () => {
      const history: IUndeliverableMessageHistory[] = [
        {
          reason: "FAILED",
          ts: 1610475430221,
          type: "UNDELIVERABLE",
        },
      ];

      expect(() => shouldMarkUndeliverable(history, [])).not.toThrow();
    });
  });
});
