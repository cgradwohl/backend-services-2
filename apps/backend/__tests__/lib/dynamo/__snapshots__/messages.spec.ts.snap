// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/messages v3 markUndeliverable should mark a message as undeliverable 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":messageStatus": "UNDELIVERABLE",
          ":now": 1632426903638,
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus, sent = :now",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 markUndeliverable should mark a message as undeliverable with configuration 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":configuration": "MOCK_CONFIGURATION",
          ":messageStatus": "UNDELIVERABLE",
          ":now": 1632426903638,
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus, sent = :now, configuration = :configuration",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 markUndeliverable should mark a message as undeliverable with internal courier error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":errorType": "INTERNAL_COURIER_ERROR",
          ":messageStatus": "UNDELIVERABLE",
          ":now": 1632426903638,
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus, sent = :now, errorType = :errorType",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 markUndeliverable should mark a message as undeliverable with provider 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":messageStatus": "UNDELIVERABLE",
          ":now": 1632426903638,
          ":provider": "MOCK_PROVIDER",
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus, sent = :now, provider = :provider",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 markUndeliverableFromDelivery should mark a message as undeliverable 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":messageStatus": "UNDELIVERABLE",
          ":now": 1632426903638,
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus, sent = :now",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 markUndeliverableFromDelivery should mark a message as undeliverable with configuration 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":configuration": "MOCK_CONFIGURATION",
          ":messageStatus": "UNDELIVERABLE",
          ":now": 1632426903638,
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus, sent = :now, configuration = :configuration",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 markUndeliverableFromDelivery should mark a message as undeliverable with internal courier error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":errorType": "INTERNAL_COURIER_ERROR",
          ":messageStatus": "UNDELIVERABLE",
          ":now": 1632426903638,
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus, sent = :now, errorType = :errorType",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 markUndeliverableFromDelivery should mark a message as undeliverable with provider 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":messageStatus": "UNDELIVERABLE",
          ":now": 1632426903638,
          ":provider": "MOCK_PROVIDER",
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus, sent = :now, provider = :provider",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 markUnroutable should mark a message as unroutable 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":messageStatus": "UNROUTABLE",
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should create a new message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_not_exists(pk)",
        "Item": Object {
          "enqueued": 1632426903638,
          "eventId": "MOCK_EVENT_ID",
          "gsi3pk": "MOCK_TENANT_ID/1",
          "gsi3sk": "1632426903638",
          "id": "MOCK_MESSAGE_ID",
          "messageId": "MOCK_MESSAGE_ID",
          "messageStatus": "ENQUEUED",
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
          "recipientId": "MOCK_RECIPIENT_ID",
          "tenantId": "MOCK_TENANT_ID",
        },
        "TableName": "jest-mock-table-name-messages-v3",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should increment the error count on a message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeNames": Object {
          "#errorCount": "errorCount",
        },
        "ExpressionAttributeValues": Object {
          ":increment": 1,
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_EVENT_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "ADD #errorCount :increment",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should mark a message as clicked 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":messageStatus": "CLICKED",
          ":timestamp": 1632427827545,
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus, clicked = :timestamp",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should mark a message as delivered 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk) AND messageStatus <> :clickedMessageStatus AND messageStatus <> :openedMessageStatus",
        "ExpressionAttributeValues": Object {
          ":clickedMessageStatus": "CLICKED",
          ":configuration": "MOCK_CONFIGURATION",
          ":messageStatus": "DELIVERED",
          ":now": 1632426903638,
          ":openedMessageStatus": "OPENED",
          ":provider": "MOCK_PROVIDER",
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "set messageStatus = :messageStatus, delivered = :now, provider = :provider, configuration = :configuration",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should mark a message as opened 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk) AND messageStatus <> :clickedMessageStatus",
        "ExpressionAttributeValues": Object {
          ":clickedMessageStatus": "CLICKED",
          ":messageStatus": "OPENED",
          ":timestamp": 1632427827545,
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus, opened = :timestamp",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should mark a message as read 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":now": 1632426903638,
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET readTimestamp = :now",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should mark a message as sent 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":configuration": "MOCK_CONFIGURATION",
          ":now": 1632426903638,
          ":provider": "MOCK_PROVIDER",
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET sent = :now, provider = :provider, configuration = :configuration",
      },
    ],
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk) AND messageStatus <> :deliveredMessageStatus",
        "ExpressionAttributeValues": Object {
          ":deliveredMessageStatus": "DELIVERED",
          ":messageStatus": "SENT",
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should mark a message as simulated 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":configuration": "MOCK_CONFIGURATION",
          ":messageStatus": "SIMULATED",
          ":now": 1632426903638,
          ":provider": "MOCK_PROVIDER",
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus, sent = :now, provider = :provider, configuration = :configuration",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should mark a message as unmapped 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":messageStatus": "UNMAPPED",
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET messageStatus = :messageStatus",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should mark a message as unread 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_MESSAGE_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "REMOVE readTimestamp",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should set the billing units id value on a message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk) AND attribute_not_exists(billed_units)",
        "ExpressionAttributeNames": Object {
          "#billed_units": "billed_units",
        },
        "ExpressionAttributeValues": Object {
          ":billed_units": 1,
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_EVENT_ID",
        },
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET #billed_units = :billed_units",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should set the email address value on a message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":email": "email@example.com",
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_EVENT_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET recipientEmail = :email",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`lib/messages v3 should set the notification id value on a message 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "ConditionExpression": "attribute_exists(pk)",
        "ExpressionAttributeValues": Object {
          ":notificationId": "MOCK_NOTIFICATION_ID",
        },
        "Key": Object {
          "pk": "MOCK_TENANT_ID/MOCK_EVENT_ID",
        },
        "ReturnValues": "NONE",
        "TableName": "jest-mock-table-name-messages-v3",
        "UpdateExpression": "SET notificationId = :notificationId",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
